### ISSUE TITLE
Tracking Issue for `new_range_api` (part of RFC 3550)

### ISSUE THREAD

**Issue by pitaj** [rust-lang/rust#125687]
Feature gate: `#![feature(new_range_api)]`

This is a tracking issue for the library additions that are part of RFC 3550: New range types

- rust-lang/rfcs#3550

Tracking issue for the language change: #123741

This feature introduces new types implementing `Copy` and `IntoIterator` that are meant to replace the legacy range types, which are `!Copy` and implement `Iterator` directly.

### Public API

```rust
// core::range

#[derive(Clone, Copy, Default, PartialEq, Eq, Hash)]
pub struct Range<Idx> {
    pub start: Idx,
    pub end: Idx,
}
#[derive(Clone, Copy, Default, PartialEq, Eq, Hash)]
pub struct RangeInclusive<Idx> {
    pub start: Idx,
    pub last: Idx, // formerly end
}
#[derive(Clone, Copy, Default, PartialEq, Eq, Hash)]
pub struct RangeFrom<Idx> {
    pub start: Idx,
}
#[derive(Copy, Clone, PartialEq, Eq, Hash)]
pub struct RangeToInclusive<Idx> {
    pub last: Idx,
}

pub struct IterRange<Idx> { <private fields> }
pub struct IterRangeInclusive { <private fields> }
pub struct IterRangeFrom { <private fields> }

impl<Idx: Step> Iterator for IterRange<Idx> { ... }
impl<Idx: Step> Iterator for IterRangeInclusive<Idx> { ... }
impl<Idx: Step> Iterator for IterRangeFrom<Idx> { ... }

impl<Idx: Step> IntoIterator for Range<Idx> {
    type IntoIter = IterRange<Idx>;
    ...
}
impl<Idx: Step> IntoIterator for RangeInclusive<Idx> {
    type IntoIter = IterRangeInclusive<Idx>;
    ...
}
impl<Idx: Step> IntoIterator for RangeFrom<Idx> {
    type IntoIter = IterRangeFrom<Idx>;
    ...
}
```

See the RFC for more details.

### Steps / History

- [x] Implementation: #125751
- [ ] [Inclusive ranges: rename `end` field to `last`](https://github.com/rust-lang/libs-team/issues/511)
- [ ] Optimize implementation
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- [ ] Figure out a way to have `IterRangeFrom` overflow panic only after yielding the maximum value, without impacting release mode
- [ ] How can we / should we optimize `IterRangeInclusive` now that it is not bound by the API constraints of legacy `RangeInclusive`?
- [ ] The set of inherent methods copied from `Iterator` present on the new range types (#125589)
- [x] The exact module paths and type names
  + Should the new types live at `std::ops::range::` instead?
  + `IterRange`, `IterRangeInclusive` or just `Iter`, `IterInclusive`? Or `RangeIter`, `RangeInclusiveIter`, ...? https://github.com/rust-lang/rust/pull/149547
- [x] Should other range-related items (like `RangeBounds`) also be moved under the `range` module?
- [x] Should `RangeFrom` even implement `IntoIterator`, or should it require an explicit `.iter()` call? Using it as an iterator [can be a footgun](https://github.com/rust-lang/libs-team/issues/304), usually people want `start..=MAX` instead. Also, it is inconsistent with `RangeTo`, which doesn't implement `IntoIterator` either. But, it's extremely useful for `it.zip(1..)`
- [ ] Should there be a way to get an iterator that modifies the range in place, rather than taking the range by value? That would allow things like `range.by_ref().next()`.
- [x] Should there be an infallible conversion from legacy to new `RangeInclusive`?
```rust
impl<Idx> From<legacy::RangeInclusive<Idx>> for RangeInclusive<Idx> {
    // How do we handle the `exhausted` case?
}
```

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by joshtriplett** [comment]
Capturing something from a verbal discussion:

In theory, since the new range types aren't iterators directly, and just get turned *into* iterators, we *could* automatically handle `for i in 10..0` and do the expected thing. However, @eholk pointed out that while that'd be intuitive for integer literals, it'd be *surprising* for the general `m..n` to implicitly do reverse iteration. So, we shouldn't do this.

**Comment by joshtriplett** [comment]
> Should `RangeFrom` even implement `IntoIterator`, or should it require an explicit `.iter()` call? Using it as an iterator [can be a footgun](https://github.com/rust-lang/libs-team/issues/304), usually people want `start..=MAX` instead. Also, it is inconsistent with `RangeTo`, which doesn't implement `IntoIterator` either. But, it's extremely useful for `it.zip(1..)`.

We discussed this in today's @rust-lang/libs-api meeting, and our conclusion was: the usefulness of `.zip(n..)` is high enough that we should make this work. We should ensure that it yields values up to the maximum integer value, and then panics if asked for another value.

**Comment by Amanieu** [comment]
We discussed the unresolved questions in the libs-api meeting yesterday.

> * The set of inherent methods copied from `Iterator` present on the new range types ([Decide which methods to add to new `Range` types #125589](https://github.com/rust-lang/rust/issues/125589))

There was a lot of opposition to inherent methods, especially `map` and `rev`, because they would need to behave differently on a value type than on an iterator. `map` would be expected to work like `Option::map` and apply a function to the start and end bounds. `rev` would be expected to return a range (not an iterator).

However there was some support for adding an `iter` inherent method which would be a short-hand for `.clone().into_iter()`. We felt that the additional 5 characters of `into_iter` (especially the underscore) would make people's code too verbose for such a common operation.

> * The exact module paths and type names
>   
>   * Should the new types live at `std::ops::range::` instead?
>   * `IterRange`, `IterRangeInclusive` or just `Iter`, `IterInclusive`? Or `RangeIter`, `RangeInclusiveIter`, ...?
> 
> * Should other range-related items (like `RangeBounds`) also be moved under the `range` module?

We were mostly happy with how things are currently set out in the PR.

> * Should `RangeFrom` even implement `IntoIterator`, or should it require an explicit `.iter()` call? Using it as an iterator [can be a footgun](https://github.com/rust-lang/libs-team/issues/304), usually people want `start..=MAX` instead. Also, it is inconsistent with `RangeTo`, which doesn't implement `IntoIterator` either. But, it's extremely useful for `it.zip(1..)`

It should implement `IntoIterator` since, as mentioned, it is useful for things like `it.zip(1..)`.

We also discussed the `RangeFrom` iterator: the implementation should add a hidden `bool` so that any overflow panics happen *after* returning the maximum integer value, not before. To avoid issues with LLVM's loop optimizations, we may want this `bool` to only be used in debug builds: in release builds it may acceptable to allow the iterator to wrap.

> * Should there be a way to get an iterator that modifies the range in place, rather than taking the range by value? That would allow things like `range.by_ref().next()`.

This should be deferred for a future API proposal, only if it turns out there is a need for this.

> * Should there be an infallible conversion from legacy to new `RangeInclusive`?

Yes there should be. The case of converting an exhausted `RangeInclusive` iterator should be extremely rare in practice, so it's fine to panic in that case.

**Comment by pitaj** [comment]
> There was a lot of opposition to inherent methods, especially `map` and `rev`, because they would need to behave differently on a value type than on an iterator.

I understand the arguments for purity here, but the immense utility these offer should be taken into account. I'll remove them from the PR, but I think the question should be reconsidered before the edition change is stabilized.

<details><summary>Arguments for `map` and `rev`</summary>

I think range types are in a very unique position, here. `map` and `rev` are
1. commonly used in the wild
2. present in much documentation
3. demonstrated in countless learning materials
4. a useful entry into iterator chains

While shaving five characters and the shift necessary for the underscore is helpful, the remaining seven characters and two shifts for `.iter()` is still a burden and needless visual noise.

My position is that there is no other meaning we could assign to `map` that would be acceptable given the history of using it in this position.

### `map` is expected to map the bounds, not the yielded elements

1. It is already a widely used pattern
2. Changing `map` to mean "map the bounds" would be a sizeable hazard, because it could silently change the meaning of `for (1..5).map(|n| n*2) { ... }` across an edition boundary.
3. So, a function doing so would need to be named something like `map_bounds` anyways.
4. `Range` is not like `Option` or `Vec`. It's not commonly used to hold values, it's purpose is to generate values.

### `rev` is expected to return something like `RevRange`

IMO this argument is stronger, because `for x in (1..5).rev()` would still work. That covers one of the most common use-cases. However, I don't see much use for a `RevRange` type.

What would it implement besides `IntoIterator`?
- not `RangeBounds`
- probably not `Index`
- could have `is_empty` and `contains` but those are direction-agnostic

It it only serves as an iterable, we may as well just use `Rev<IterRange<_>>` anyways.

</details>

**Comment by dvdsk** [comment]
I wish we had a `map` transforming the bounds of a Range. Though I get that we can not get it since it would get really confusing if the range types implemented `IntoIterator`. 

Therefore I really like the idea of a `map_bounds` function. Re-creating a range simply to map the bounds takes a lot of space for a simple thing. That tents to get me out of the flow of reading any involved algorithm.

**Comment by balt-dev** [comment]
Thought: What's stopping anyone from making this an `enum` over `Range`, `RangeFull`, `RangeInclusive`, etc?

**Comment by sendittothenewts** [comment]
At present, the `end` member of `RangeInclusive` and `RangeToInclusive` has the same name but slightly different semantics to the `end` member of `Range` and `RangeTo`.  This strikes me as a bit of a footgun, as a seemingly-innocuous change like replacing `..=n-1` with `..n` can introduce off-by-one errors (see https://github.com/rust-lang/rust-clippy/issues/3307#issuecomment-2119369210).  I suggest renaming the `end` member of `RangeInclusive` and `RangeToInclusive`, perhaps to `last`, consistent with the `IterRangeInclusive::last` method.

**Comment by pitaj** [comment]
It's a good idea but we probably don't want to break code using `RangeToInclusive` just for this. 

[Open an ACP](https://github.com/rust-lang/libs-team/issues/new?template=api-change-proposal.md) to discuss.

**Comment by PetrGlad** [comment]
I am in favor of using any `IntoIterator` type in `for`, so I do not understand why
> it'd be _surprising_ for the general `m..n` to implicitly do reverse iteration.
@joshtriplett can you give an example where this is a problem? 
I can imagine If a bound type prefers to iterate in an unusual way, it is responsibility of its author. I do not think that for syntax is a place that should gate-keep it...

10..0 is empty, is this what you meant?

**Comment by bryceberger** [comment]
Was there discussion about having a `.len()` inherent method? I didn't see it mentioned in #125589. The old range types implement it for numeric types through `ExactSizeIterator`, and it seems strange to rewrite `range.len()` into `range.iter().len()`.

**Comment by pitaj** [comment]
I'd love to put a `width` function on the range types.

For `Range` this is super easy. You can just return `T`: 0 if empty, `end-start` otherwise.

For `RangeInclusive` this is more complicated, though. You want to return `end-start+1` in the non-empty case, but that can overflow `T`. And in fact, a couple of the `ExactSizeIterator` impls for the old RangeInclusive are incorrect for this reason.

There are a couple options I've thought of:

1. Return `Option<T>`: 0 if empty, `None` on overflow, `Some(end-start+1)` otherwise
2. Return `PlusOne<T>`: this type implements `TryInto<T>`, `TryInto<usize>`, `TryInto<u128>`, etc. And those impls are the only way to get a value out.

**Comment by programmerjake** [comment]
> [Open an ACP](https://github.com/rust-lang/libs-team/issues/new?template=api-change-proposal.md) to discuss.

since [the ACP was accepted](https://github.com/rust-lang/libs-team/issues/511#issuecomment-2576358093), can implementing it be added as a blocking issue here?

**Comment by tgross35** [comment]
> * The exact module paths and type names:
>   * `IterRange`, `IterRangeInclusive` or just `Iter`, `IterInclusive`? Or `RangeIter`, `RangeInclusiveIter`, ...?

I think `Iter` may be better as a suffix than a prefix, so would like to suggest renaming to `RangeIter` and similar. For common combinations like `IntoIter`/`IterMut`, `RangIterMut` and `RangeIntoIter` read much better than `IterMutRange` or `IntoIterRange`.

Obviously those types aren't meaningful but it seems like a useful litmus test, since the precedent here will probably wind up used elsewhere.

**Comment by tgross35** [comment]
> My thought process was this: typically the type would just be called `Iter` (like `slice::Iter`). The type of range is just a differentiator, so should come second.

Being used to reading `slice::Iter` is probably part of why I would name a reexport `SliceIter` rather than `IterSlice` :)

I think there is also some ecosystem convention that `FooIter` is an iterator over a type `Foo`, vs. `IterFoo` is an iterator yielding `Foo`s. Examples are `hashbrown`'s `IterBuckets` (iterator on a hashtable yielding buckets), `regex::SetMatchesIter` (iterator over a `SetMatches` struct yielding matches), `itertools::structs::RcIter` (iterator over an `Rc` yielding its items), so `RangeIter` would fit into this. Bit of a weak convention though.

**Comment by traviscross** [comment]
We talked about this in the lang call today.  We were all excited to see this go forward.  We appointed @Amanieu, in his role as a lang-advisor, as the champion for the language feature.

**Comment by sendittothenewts** [comment]
Can I suggest that the `core::range::RangeBounds` alias not be stabilized until `IntoBounds` stabilizes?  As I suggested in https://github.com/rust-lang/rust/issues/136903#issuecomment-3498341296, `IntoBounds` might be better as a supertrait of `RangeBounds` instead of a subtrait, and stabilizing `core::range::RangeBounds` would kind of close the door on that.  This need not be a blocker for stabilizing the new types though.

**Comment by pitaj** [comment]
That was handled by #148703

**Comment by Kile-Asmussen** [comment]
Probably a stupid question, but why is there no `RangeTo<Idx>` and `RangeFull` in the OP? Just an oversight or technical limitation?
